buffer/src/main/java/io/netty/buffer/api/CompositeBuffer.java

```diff
@@ -0,0 +1,2004 @@
+        }
+        if (minimumGrowth < 0) {
+            throw new IllegalArgumentException("The minimum growth cannot be negative: " + minimumGrowth + '.');
+        }
+        if (readOnly) {
+            throw bufferIsReadOnly(this);
+        }
+        if (writableBytes() >= size) {
+            // We already have enough space.
+            return;
+        }
+
+        if (allowCompaction && size <= roff) {
+            // Let's see if we can solve some or all of the requested size with compaction.
+            // We always compact as much as is possible, regardless of size. This amortizes our work.
+            int compactableBuffers = 0;
+            for (Buffer buf : bufs) {
+                if (buf.capacity() != buf.readerOffset()) {
+                    break;
+                }
+                compactableBuffers++;
+            }
+            if (compactableBuffers > 0) {
+                Buffer[] compactable;
+                if (compactableBuffers < bufs.length) {
+                    compactable = new Buffer[compactableBuffers];
+                    System.arraycopy(bufs, 0, compactable, 0, compactable.length);
+                    System.arraycopy(bufs, compactable.length, bufs, 0, bufs.length - compactable.length);
+                    System.arraycopy(compactable, 0, bufs, bufs.length - compactable.length, compactable.length);
+                } else {
+                    compactable = bufs;
+                }
+                for (Buffer buf : compactable) {
+                    buf.resetOffsets();
+                }
+                computeBufferOffsets();
+                if (writableBytes() >= size) {
+                    // Now we have enough space.
+                    return;
+                }
+            } else if (bufs.length == 1) {
+                // If we only have a single component buffer, then we can safely compact that in-place.
+                bufs[0].compact();
+                computeBufferOffsets();
+                if (writableBytes() >= size) {
+                    // Now we have enough space.
+                    return;
+                }
+            }
+        }
+
+        int growth = Math.max(size - writableBytes(), minimumGrowth);
+        BufferAllocator.checkSize(capacity() + (long) growth);
+        Buffer extension = allocator.allocate(growth, order());
+        unsafeExtendWith(extension);
+    }
+
+    /**
+     * Extend this composite buffer with the given extension buffer.
+     * This works as if the extension had originally been included at the end of the list of constituent buffers when
+     * the composite buffer was created.
+     * The extension buffer is added to the end of this composite buffer, which is modified in-place.
+     *
+     * @see #compose(BufferAllocator, Send...)
+     * @param extension The buffer to extend the composite buffer with.
+     */
+    public void extendWith(Send<Buffer> extension) {
```

Comment:
Should we return `this` here to make it more consistent with `Buffer` and allow method chaining ?